evalml/pipelines/components/transformers/encoders/ordinal_encoder.py

```diff
@@ -0,0 +1,320 @@
+                    "Categories argument must contain a list of categories for each ordinal feature",
+                )
+
+            # Categories, as they're passed into SKOrdinalEncoder should be in the same order
+            # as the data's Ordinal.order categories even if it's a subset
+            categories = []
+            for i, col_categories in enumerate(input_categories):
+                categories_order = ww_logical_types[self.features_to_encode[i]].order
+
+                ordered_categories = [
+                    cat for cat in categories_order if cat in col_categories
+                ]
+                categories.append(ordered_categories)
+        else:
+            # Categories unspecified - use ordered categories from a columns' Ordinal logical type
+            # and account for top_n
+            categories = []
+            for col in X_t[self.features_to_encode]:
+                ltype = ww_logical_types[col]
+                # Copy the order list, since we might mutate it later by adding nans
+                # and don't want to impact the Woodwork types
+                column_ordered_categories = ltype.order.copy()
+
+                if top_n is None or len(column_ordered_categories) <= top_n:
+                    unique_values = column_ordered_categories
+                else:
+                    # Don't count nans as a value
+                    value_counts = X_t[col].value_counts(dropna=True).to_frame()
+                    value_counts = value_counts.sample(
+                        frac=1,
+                        random_state=self._initial_state,
+                    )
+                    value_counts = value_counts.sort_values(
+                        [col],
+                        ascending=False,
+                        kind="mergesort",
+                    )
+                    unique_values = value_counts.head(top_n).index.tolist()
+                    # Categories should be in the same order as the data's Ordinal.order categories
+                    # instead of sorted by most frequent
+                    unique_values = [
+                        cat for cat in column_ordered_categories if cat in unique_values
+                    ]
+
+                categories.append(unique_values)
+
+        # Add any null values into the categories lists so that they aren't treated as unknown values
+        # This is needed because Ordinal.order won't indicate if nulls are present, and SKOrdinalEncoder
+        # requires any null values be present in the categories list if they are to be encoded as
+        # missing values
+        if isinstance(categories, list):
+            for i, col in enumerate(X_t[self.features_to_encode]):
+                if X_t[col].isna().any():
+                    categories[i] += [np.nan]
+
+        encoded_missing_value = self.parameters["encoded_missing_value"]
+        if encoded_missing_value is None:
+            encoded_missing_value = np.nan
```

Comment:
I'm having trouble figuring out if this section is worth the effort or not. If I'm following correctly, we should only allow `np.nan` as a category if `handle_unknown` is `"use_encoded_value"`. In that case, we don't allow `encoded_missing_value` to be None (lines 86 and 87 in init), so the second half of this seems moot.

In that case (I've been thinking while typing here), I might replace this section (lines 220-227) with:
```
if self.parameters["handle_unknown"] == "use_encoded_value":
    for i, col in enumerate(X_t[self.features_to_encode]):
        if X_t[col].isna().any():
            categories[i].append(np.nan)
```
Then, when instantiating the encoder, I'd just do `encoded_missing_value=self.parameters["encoded_missing_value"] or np.nan`. 
I think that would be a clearer way to do it, as it's a little confusing as it is now.

Reply:
(I haven't totally thought through what you're actually talking about in this comment, but the use of `or` here brought up thoughts I didn't want to forget!)

Whenever I see the `or` operator used like this, I always worry about falsy values falling to the second option, and I think we might be in the clear here on a technicality, but I'm not sure we'd want to still go with it. 

The encoded missing value has to be numeric in nature, which takes away empty string, empty set, etc as issues. `0` is where I'm not sure; technically users aren't allowed to use `0` since it'll always be the first encoded value unless none of the known categories are seen in the column. But instead of ever getting an error for using `0`, we'd convert to np.nan and be able to proceed. So we'd be allowing something that's not allowed in most cases. And in the few cases where it's allowed, we wouldn't be doing what the user asked for.

Reply:
@eccabay So "missing" value handling is actually totally independent of "unknown" handling with the way we do things. (Lines 86 and 87 are referencing `unknown_value`)

We don't ever want missing values to get counted as unknown (because with Woodwork, they can't actually be categories and can only ever be missing values - the same thing as I mentioned in another comment - so I wouldn't consider them unknown if they're outside of the concept of categories to woodwork). If we want the option to error when nans are seen, I think that should be its own parameter. 

And I think my reservations about using the `or` syntax still hold, so I'm going to keep it as is for now. 

But overall, I agree with the sentiment that it feels like there should be a cleaner way to handle a lot of this logic. Maybe fresh eyes next week will see something I'm missing right now!

Reply:
Ah, thanks for pointing out the missing vs unknown distinction, I totally combined them! Your point about `or` is also definitely important, you're right that we should avoid it!